### PR TITLE
Add tutorial to demonstrate how to train a generative model that can generate large image volumes

### LINKED ISSUE #1744: Adding tutorial to demonstrate how to train a generative model that can generate large image volumes
Add tutorial to demonstrate how to train a generative model that can generate large image volumes (e.g., of size 512 x 512 x 512). [#1745]

### PR DESCRIPTION
**Pull request by dongyang0122**
Fixes #1744.

### Description
Add tutorial to demonstrate how to train a generative model that can generate large image volumes.

### Checks

- [ ] Avoid including large-size files in the PR.
- [ ] Clean up long text outputs from code cells in the notebook.
- [ ] For security purposes, please check the contents and remove any sensitive info such as user names and private key.
- [ ] Ensure (1) hyperlinks and markdown anchors are working (2) use relative paths for tutorial repo files (3) put figure and graphs in the `./figure` folder
- [ ] Notebook runs automatically `./runner.sh -t <path to .ipynb file>`

### CODE REVIEW

File: 3d_generation/maisi/scripts/infer_using_diff_model.py [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+            ).to("cuda")
+            z = outputs
+
+            _temp = recon1(
+                z[
+                    ...,
+                    : target_shape[0] // 4 // 4 * 3,
+                    : target_shape[1] // 4 // 4 * 3,
+                    : target_shape[2] // 4 // 4 * 3,
+                ],
```
Comment (ericspod) [inline 3d_generation/maisi/scripts/infer_using_diff_model.py:334]:
```suggestion
                z[
                    ...,
                    : target_shape[0] // 4 // 4 * 3,
                    : target_shape[1] // 4 // 4 * 3,
                    : target_shape[2] // 4 // 4 * 3
                ],
```

All of these strange formattings are caused by superfluous commas. I've seen this elsewhere and it's an odd thing black does.

File: 3d_generation/maisi/scripts/infer_using_diff_model.py [hunk 2]
```diff
@@ -0,0 +1,586 @@
+#!/usr/bin/env python
```
Comment (ericspod) [inline 3d_generation/maisi/scripts/infer_using_diff_model.py:1]: I think all the source files need license headers?

File: generative/maisi/docs/diff_model_explaination.md [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    np.random.seed(0)
+    set_determinism(seed=0)
+    ```
+
+2. **Define Paths**:
+    ```python
+    dataroot = "/mnt/drive2/data_128"
+    filenames_filepath = "/mnt/drive2/data_128/filenames_image_nii_autopet.txt"
+    output_root_embedding = "/mnt/drive2/encoding_128"
+    autoencoder_root = "/workspace/monai/generative/from_canz"
```
Comment (mingxin-zheng) [inline generative/maisi/docs/diff_model_explaination.md:28]: Can we remove the personal info in this path? Thanks!

File: generative/maisi/docs/diff_model_explaination.md [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    ```
+
+3. **Distributed Initialization**:
+    ```python
+    dist.init_process_group(backend="nccl", init_method="env://")
```
Comment (mingxin-zheng) [inline generative/maisi/docs/diff_model_explaination.md:33]: I would suggest to add an import here for the first time declaration
```
import torch.distributed as dist
```

**Comment by mingxin-zheng** [comment]
Hi @dongyang0122 , some signature is missing:
https://github.com/Project-MONAI/tutorials/pull/1745/checks?check_run_id=27169416504

```
Summary
Commit sha: [f159834](https://github.com/Project-MONAI/tutorials/pull/1745/commits/f1598348036b7d64267f1c7a8f1fa76d4a5d37aa), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [50d0382](https://github.com/Project-MONAI/tutorials/pull/1745/commits/50d0382c5194f2eb185d0a7dceea148eaec406da), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [e818c46](https://github.com/Project-MONAI/tutorials/pull/1745/commits/e818c46d56d40cca03fdaca1d7db82577fccf172), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [e033503](https://github.com/Project-MONAI/tutorials/pull/1745/commits/e03350303104d65254f84ca482b752ac83f1de48), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [e41d619](https://github.com/Project-MONAI/tutorials/pull/1745/commits/e41d6197c989de35751f53b0a662d9f0a7ef344c), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [d78b35a](https://github.com/Project-MONAI/tutorials/pull/1745/commits/d78b35a1f230319fb898b68c41ccabe29b1d9c27), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [a5de734](https://github.com/Project-MONAI/tutorials/pull/1745/commits/a5de734e08ffc74e5bada4b4cef5e2f62c2d02cf), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [79d620f](https://github.com/Project-MONAI/tutorials/pull/1745/commits/79d620f0210281b7d09a1827cfe88b9606f9ad4d), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [74f69a4](https://github.com/Project-MONAI/tutorials/pull/1745/commits/74f69a4d34ff116c498913c98c9af135f62bc47b), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [3268421](https://github.com/Project-MONAI/tutorials/pull/1745/commits/32684213f89da42fdef71b3709948693a693bc59), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.
Commit sha: [5847ed8](https://github.com/Project-MONAI/tutorials/pull/1745/commits/5847ed823621d4a3f970058ae7b1c611a9f4c5d6), Author: Dong Yang, Committer: Dong Yang; The sign-off is missing.

```

File: 3d_generation/maisi/scripts/infer_using_diff_model.py (hunk 1 above)
Comment (dongyang0122) [reply 3d_generation/maisi/scripts/infer_using_diff_model.py]: Hi @ericspod, I have updated the script. Please proceed. Thanks.

File: 3d_generation/maisi/scripts/infer_using_diff_model.py (hunk 2 above)
Comment (dongyang0122) [reply 3d_generation/maisi/scripts/infer_using_diff_model.py]: Hi @ericspod, I have updated the script. Please proceed. Thanks.

File: generative/maisi/docs/diff_model_explaination.md (hunk 3 above)
Comment (dongyang0122) [reply generative/maisi/docs/diff_model_explaination.md]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

File: generative/maisi/docs/diff_model_explaination.md (hunk 4 above)
Comment (dongyang0122) [reply generative/maisi/docs/diff_model_explaination.md]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

**Comment by KumoLiu** [comment]
Hi @dongyang0122, could you please take a look at the ci issue? You can see the error using details. Thanks!

File: generative/maisi/scripts/diff_model_create_training_data.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+import monai
+from monai.transforms import Compose
+from monai.utils import set_determinism
+
+from utils import define_instance, load_autoencoder_ckpt
+
+# Set the random seed for reproducibility
+torch.manual_seed(0)
+np.random.seed(0)
+set_determinism(seed=0)
```
Comment (KumoLiu) [inline generative/maisi/scripts/diff_model_create_training_data.py:32]:
```suggestion
set_determinism(seed=0)
```

`set_determinism` already include the above two lines, we can remove them.

File: generative/maisi/maisi_diff_unet_training_tutorial.ipynb [hunk 6]
```diff
@@ -0,0 +1,475 @@
+{
```
Comment (KumoLiu) [inline generative/maisi/maisi_diff_unet_training_tutorial.ipynb:1]: <p>I guess "To provide the "Hello World!" experience of Auto3D in this notebook" here should be updated for generative?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1745/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

File: generative/maisi/scripts/diff_model_create_training_data.py (hunk 5 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_create_training_data.py]: Hi @KumoLiu, I have updated the script. Please proceed. Thanks.

File: generative/maisi/maisi_diff_unet_training_tutorial.ipynb (hunk 6 above)
Comment (dongyang0122) [reply generative/maisi/maisi_diff_unet_training_tutorial.ipynb:1]: Hi @KumoLiu, I have updated the notebook. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_create_training_data.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+import argparse
+import json
+import os
+from pathlib import Path
+
+import nibabel as nib
+import numpy as np
+import torch
+import torch.distributed as dist
+import torch.nn.functional as F
```
[the remark this answers is not in the document]
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_create_training_data.py]: Hi @KumoLiu, I have updated the script. Please proceed. Thanks.

File: generative/maisi/maisi_diff_unet_training_tutorial.ipynb [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
```
Comment (KumoLiu) [inline generative/maisi/maisi_diff_unet_training_tutorial.ipynb:1]: <p>Also need install xformers to run this notebook.</p><p><code></code><code></p><p>!python -c "import xformers" || pip install -q xformers --index-url <a href="https://download.pytorch.org/whl/cu121" target="_blank">https://download.pytorch.org/whl/cu121</a></p><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1745/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: generative/maisi/scripts/diff_model_infer.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+from datetime import datetime, timedelta
+from tqdm import tqdm
+
+from monai.inferers import sliding_window_inference
+from monai.utils import set_determinism
+
+from utils import define_instance, load_autoencoder_ckpt
+
+
+class ReconModel(torch.nn.Module):
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_infer.py:30]: There's a ReconModel under `generative/maisi/scripts/sample.py`. The class seems the same as the one here.

File: generative/maisi/scripts/diff_model_infer.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    set_determinism(random_seed)
+    print(f"random seed: {random_seed}")
+
+    output_size = tuple(args.diffusion_unet_inference["dim"])
+    out_spacing = tuple(args.diffusion_unet_inference["spacing"])
+
+    ckpt_filepath = f"{args.model_dir}/{args.model_filename}"
+    output_prefix = args.output_prefix
+
+    if local_rank == 0:
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_infer.py:93]: Maybe we use the monai RankFilter and the logging module to replace all the `print` in this function.

File: generative/maisi/scripts/diff_model_infer.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
+        output_path = f"{args.output_dir}/{output_prefix}_seed{random_seed}_size{output_size[0]:d}x{output_size[1]:d}x{output_size[2]:d}_spacing{out_spacing[0]:.2f}x{out_spacing[1]:.2f}x{out_spacing[2]:.2f}_{timestamp}.nii.gz"
+        os.makedirs(os.path.dirname(output_path), exist_ok=True)
+        nib.save(new_image, output_path)
+        print(f"Saved {output_path}.")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Diffusion Model Inference")
+    parser.add_argument("--env_config", type=str, required=True, help="Path to environment configuration file")
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_infer.py:217]: There are multiple environment config in the folder `generative/maisi/configs` now. Can you provide an example filename to help the user to find out the right file among all the JSON files? Thanks!

File: generative/maisi/scripts/diff_model_infer.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    parser.add_argument("--model_config", type=str, required=True, help="Path to model configuration file")
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_infer.py:218]: Same as above

File: generative/maisi/scripts/diff_model_train.py [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+
+    dist.init_process_group(backend="nccl", init_method="env://", timeout=timedelta(seconds=7200))
+
+    local_rank = int(os.environ["LOCAL_RANK"])
+    world_size = int(dist.get_world_size())
+    device = torch.device("cuda", local_rank)
+    torch.cuda.set_device(device)
+    print(f"Using {device} of {world_size}")
+
+    if local_rank == 0:
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_train.py:77]: Same comment as above on the use of `RankFilter` and logging module to replace the `print`

File: generative/maisi/scripts/diff_model_train.py [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+    optimizer = torch.optim.Adam(params=unet.parameters(), lr=lr)
+    print(f"optimizer -> Adam; lr -> {lr}.")
+
+    total_steps = (num_epochs * len(train_loader.dataset)) / num_images_per_batch
+    print(f"total number of training steps: {total_steps}.")
+
+    lr_scheduler = torch.optim.lr_scheduler.PolynomialLR(optimizer, total_iters=total_steps, power=2.0)
+
+    loss_pt = torch.nn.L1Loss()
+    best_loss = 1e4
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_train.py:191]: This variable seems unused.

File: generative/maisi/scripts/diff_model_train.py [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+                    "scale_factor": scale_factor,
+                    "unet_state_dict": unet_state_dict,
+                },
+                f"{ckpt_folder}/{args.model_filename}",
+            )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Diffusion Model Training")
+    parser.add_argument("--env_config", type=str, required=True, help="Path to environment configuration file")
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_train.py:287]: Same as the comment above.

File: generative/maisi/scripts/diff_model_infer.py [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+    print(f"scale_factor -> {scale_factor}.")
+
+    autoencoder.eval()
+    unet.eval()
+
+    recon_model = ReconModel(autoencoder=autoencoder, scale_factor=scale_factor).to(device)
+
+    with torch.no_grad():
+        with torch.cuda.amp.autocast(enabled=True):
+            noise = torch.randn(
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_infer.py:140]: The use of `torch.randn((t,)).to(device)` could be simplified to `torch.randn([t], device=device)`.

File: generative/maisi/scripts/diff_model_infer.py [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+        super().__init__()
+        self.autoencoder = autoencoder
+        self.scale_factor = scale_factor
+
+    def forward(self, z):
+        recon_pt_nda = self.autoencoder.decode_stage_2_outputs(z / self.scale_factor)
+        return recon_pt_nda
+
+
+def diff_model_infer(env_config_path: str, model_config_path: str):
```
Comment (mingxin-zheng) [inline generative/maisi/scripts/diff_model_infer.py:41]: Overall I would suggest to break this function into smaller sub-function, which may improve the readability. Thanks!

File: generative/maisi/maisi_diff_unet_training_tutorial.ipynb (hunk 8 above)
Comment (dongyang0122) [reply generative/maisi/maisi_diff_unet_training_tutorial.ipynb:1]: Hi @KumoLiu, I have updated the notebook. Please proceed. Thanks.

**Comment by mingxin-zheng** [comment]
Shall we add some doc to explain how/where to run the diff-model_{infer/train}.py later (since the markdown is removed from the PR)

File: generative/maisi/scripts/diff_model_infer.py (hunk 9 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_infer.py]: Hi @mingxin-zheng, I have updated the notebook. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_train.py (hunk 14 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_train.py]: Hi @mingxin-zheng, I have updated the notebook. Please proceed. Thanks.

**Comment by dongyang0122** [comment]
> Shall we add some doc to explain how/where to run the diff-model_{infer/train}.py later (since the markdown is removed from the PR)

There will be a README PR later. We will add this information in the unified README file.

**Comment by Can-Zhao** [comment]
The model architecture is also defined in ['config_maisi.json'](https://github.com/Project-MONAI/tutorials/blob/main/generative/maisi/configs/config_maisi.json). Do you think we can reuse it and avoid duplicate network definition?

File: generative/maisi/scripts/diff_model_infer.py (hunk 10 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_infer.py]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_infer.py (hunk 11 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_infer.py]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_infer.py (hunk 12 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_infer.py]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_train.py (hunk 13 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_train.py]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_train.py (hunk 15 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_train.py]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_create_training_data.py [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+        nda_image = nda_image.numpy().squeeze()
+        print("new", nda_image.shape, new_affine)
+
+        try:
+            out_path = Path(out_filename)
+            out_path.parent.mkdir(parents=True, exist_ok=True)
+            print(f"out_filename: {out_filename}")
+
+            with torch.cuda.amp.autocast():
+                with torch.no_grad():
```
[the remark this answers is not in the document]
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_create_training_data.py]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_infer.py (hunk 16 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_infer.py]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

File: generative/maisi/scripts/diff_model_infer.py (hunk 17 above)
Comment (dongyang0122) [reply generative/maisi/scripts/diff_model_infer.py]: Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.

**Comment by dongyang0122** [comment]
> The model architecture is also defined in ['config_maisi.json'](https://github.com/Project-MONAI/tutorials/blob/main/generative/maisi/configs/config_maisi.json). Do you think we can reuse it and avoid duplicate network definition?

Hi @Can-Zhao, I have updated the scripts to reuse model definition from `config_maisi.json`. Thanks for suggestion.

File: 3d_generation/maisi/scripts/infer_using_diff_model.py (hunk 2 above)
Comment (mingxin-zheng) [reply 3d_generation/maisi/scripts/infer_using_diff_model.py]: I saw the license headers were added. Closing this thread for now.

File: 3d_generation/maisi/scripts/infer_using_diff_model.py (hunk 1 above)
Comment (mingxin-zheng) [reply 3d_generation/maisi/scripts/infer_using_diff_model.py]: Closing this for now as it is no longer valid

File: generative/maisi/maisi_diff_unet_training_tutorial.ipynb (hunk 8 above)
Comment (mingxin-zheng) [reply generative/maisi/maisi_diff_unet_training_tutorial.ipynb:1]: Closing this for now as the line command is added.

**Review by mingxin-zheng — APPROVED** [review]
It looks good to me and I just set the DCO to pass. Hi @KumoLiu , is it necessary to push another commit by @dongyang0122 ?

**Comment by KumoLiu** [comment]
> Hi @KumoLiu , is it necessary to push another commit by @dongyang0122 ?

No, should be fine for now.

**Pull request merged**